pydicom/dicomdir.py

```diff
@@ -82,54 +105,682 @@ def __init__(self,
             """
-            sibling_list = [record]
-            current_record = record
-            while ('OffsetOfTheNextDirectoryRecord' in current_record and
-                   current_record.OffsetOfTheNextDirectoryRecord):
-                offset_of_next = current_record.OffsetOfTheNextDirectoryRecord
-                sibling = map_offset_to_record[offset_of_next]
-                sibling_list.append(sibling)
-                current_record = sibling
-            return sibling_list
-
-        # Build the mapping from file offsets to records
-        records = self.DirectoryRecordSequence
-        if not records:
-            return
-
-        map_offset_to_record = {}
-        for record in records:
-            offset = record.seq_item_tell
-            map_offset_to_record[offset] = record
-        # logging.debug("Record offsets: " + map_offset_to_record.keys())
-
-        # Find the children of each record
-        for record in records:
-            record.children = []
-            if 'OffsetOfReferencedLowerLevelDirectoryEntity' in record:
-                child_offset = (record.
-                                OffsetOfReferencedLowerLevelDirectoryEntity)
-                if child_offset:
-                    child = map_offset_to_record[child_offset]
-                    record.children = get_siblings(child, map_offset_to_record)
-
-        self.patient_records = [
-            record for record in records
-            if getattr(record, 'DirectoryRecordType') == 'PATIENT'
+            # If no children we are at the end of the branch
+            if not record.children:
+                # Build up the FileInstance by traversing back to the root
+                instance = FileInstance(self)
+                instance.add_record(record)
+                parent = record.parent
+                while parent:
+                    instance.add_record(parent)
+                    parent = parent.parent  # Move up one level
+
+                self._instances.append(instance)
+
+                return instance
+
+            # Recurse
+            branch = {}
+            for child in record.children:
+                branch[child] = recurse_branch(child)
+
+            return branch
+
+        # Top-level records
+        for record in self._dicomdir.root:
+            tree[record] = recurse_branch(record)
+
+        self._tree = tree
+
+    @property
+    def path(self) -> str:
+        """Return the absolute path to the File-set root directory as
+        :class:`str`.
+        """
+        return os.fspath(Path(self._dicomdir.filename).resolve().parent)
+
+    def as_tree(self, hierarchy: Optional[List[str]] = None) -> dict:
+        """Return a dict containing the File-set's SOP Instance hierarchy.
+
+        .. warning::
+
+            SOP Instances described using ``PRIVATE`` records are not
+            supported and will not be included in the returned tree.
+
```

Comment:
Is there a technical reason for not supporting `PRIVATE` records?  I'm not sure what is right here, but would typically go for the 'round trip' philosophy of keeping all original items.

Reply:
Yeah, I'm not happy with this. It'll be reworked with I do the writing PR so I didn't want to spend too much time on it.

Reply:
Oh, and the technical reason is; at the moment the records are kept in `FileInstance` in a `dict` using the record type as the keyword (for non PRIVATE records) or the `DirectoryRecord.key` (for PRIVATE records). This is because PRIVATE records can appear at multiple levels of an instance branch (ie can have PATIENT > PRIVATE > PRIVATE > PRIVATE) so a `FileInstance` can have multiple PRIVATE records. 

Since `as_tree` uses a simple `set` comparison when checking a `FileInstances` records, it wasn't straightforward to extend the results to PRIVATE records. This is one of the reasons why I'm not completely happy with the way the various classes interact with the record tree. I need something that lets me access records by record type (which needs to account for possibly multiple PRIVATEs), while also being able to keep them/put them in tree form.

The `key` concept gets halfway there, but that would mean I'd need to iterate over the whole tree to access a particular `FileInstance` record, which I don't like. 

I'm mulling over just doing away with storing the `FileInstance._records` directly in favour of just creating them on the fly as required by the stored instance. Which would be where the next PR comes into it.

Reply:
Okay, thanks for the explanation. I'm happy to wait for the next PR.